File: Design Patterns/Behavioral/Typescript/Mediator.ts
//Centralizes complex communication between objects.

interface IAirTrafficMediator {
    sendMessage(message: string, airplane: Airplane): void
}

class AirTrafficControl implements IAirTrafficMediator {
    private airplanes: Airplane[] = []

    registerAirplane(airplane: Airplane) { this.airplanes.push(airplane) }

    sendMessage(message: string, sender: Airplane) {
        for (const airplane of this.airplanes) {
            if (airplane !== sender) {
                airplane.receive(message)
            }
        }
    }
}


class Airplane {
    constructor(
        private mediator: IAirTrafficMediator,
        public name: string
    ) { }

    send(message: string) {
        console.log(`${this.name} sends: ${message}`)
        this.mediator.sendMessage(message, this)
    }

    receive(message: string) { console.log(`${this.name} receives: ${message}`) }
}

const tower = new AirTrafficControl()

const plane1 = new Airplane(tower, "Plane A")
const plane2 = new Airplane(tower, "Plane B")
const plane3 = new Airplane(tower, "Plane C")

tower.registerAirplane(plane1)
tower.registerAirplane(plane2)
tower.registerAirplane(plane3)

plane1.send("Requesting landing clearance")

File: C#/Design Patterns/Behavioral/Mediator.cs
using System;
using System.Collections.Generic;

interface IAirTrafficMediator
{
  void RegisterAirplane(Airplane airplane);
  void SendMessage(string message, Airplane sender);
}

class AirTrafficControl : IAirTrafficMediator
{
  private List<Airplane> airplanes = new List<Airplane>();

  public void RegisterAirplane(Airplane airplane) { this.airplanes.Add(airplane); }

  public void SendMessage(string message, Airplane sender)
  {
    this.airplanes.ForEach(airplane => { if (airplane != sender) { airplane.Receive(message); } });
  }
}

class Airplane
{
  private readonly IAirTrafficMediator mediator;
  public readonly string name;

  public Airplane(IAirTrafficMediator mediator, string name)
  {
    this.mediator = mediator;
    this.name = name;
  }

  public void Send(string message)
  {
    Console.WriteLine($"{this.name} sends: {message}");
    this.mediator.SendMessage(message, this);
  }

  public void Receive(string message) { Console.WriteLine($"{this.name} receives: {message}"); }
}
class Program
{
  public static void Main(string[] args)
  {

    AirTrafficControl tower = new AirTrafficControl();

    Airplane plane1 = new Airplane(tower, "Plane A");
    Airplane plane2 = new Airplane(tower, "Plane B");
    Airplane plane3 = new Airplane(tower, "Plane C");

    tower.RegisterAirplane(plane1);
    tower.RegisterAirplane(plane2);
    tower.RegisterAirplane(plane3);

    plane1.Send("Requesting landing clearance");
  }
}